File: ruby/lib/bool/explicit.rb
module Bool
  class Explicit
    if RUBY_PLATFORM =~ /java/
      require 'bool_ext'

      def self.new
        Java::Bool::Explicit.new
      end

    else
      def var(node, vars)
        node.name
      end

      def and(node, vars)
        "(" + explicit(node.left) + " && " + explicit(node.right) + ")"
      end

      def or(node, vars)
        "(" + explicit(node.left) + " || " + explicit(node.right) + ")"
      end

      def not(node, vars)
        "!" + explicit(node.other)
      end

      private

      def explicit(node)
        node.describe_to(self, nil)
      end
    end
  end
end

File: javascript/lib/explicit.js
module.exports = function Explicit() {
  this.var = function(var_node, vars) {
    return var_node.name;
  };

  this.and = function(and_node, vars) {
    return "(" + and_node.left.describeTo(this, vars) + " && " + and_node.right.describeTo(this, vars) + ")";
  };

  this.or = function(or_node, vars) {
    return "(" + or_node.left.describeTo(this, vars) + " || " + or_node.right.describeTo(this, vars) + ")";
  };

  this.not = function(not_node, vars) {
    return "!" + not_node.refnode.describeTo(this, vars);
  };
};
